### PR TITLE
Implement range to moc function

### LINKED ISSUE #31: Healpix Range to MOC
I am wondering if some convenience code or doc example of how to go from the healpix range sets to a MOC might be useful.

### PR DESCRIPTION
**Pull request by bparazin**
I have added a function Point.to_moc, which takes in an array-like of healpix ranges (given as half-open intervals), the healpix nside, and the indexing scheme (ring or nested) and does a bit of quick work on them to convert to the input type for MOC.from_healpix_cells. This is still WIP because I haven't thoroughly tested it.

Closes https://github.com/skyportal/healpix-alchemy/issues/31 if merged

### CODE REVIEW

**Comment by bparazin** [comment]
oh dear let's fix these pep8 issues

**Comment by bparazin** [comment]
Oh and I'm probably gonna want to write a test or 2 to make sure this works as well, right? Make codecov happy?

**Comment by lpsinger** [comment]
> Oh and I'm probably gonna want to write a test or 2, right? Make codecov happy?

Yes indeed.

**Comment by bparazin** [comment]
When it comes to writing the tests, do you have any suggestions for how to set it up? Right now I am thinking to somehow randomly generate an MOC (probably just by pulling a certain number of random UNIQ indices), decompose that into an l=29(?) healpix list (probably by seeing which l=29(?) healpix pixels intersect with said MOC), and then run it through the to_moc code to see if it produces the same MOC. Any thoughts there? 

Oh and I will probably do 2 tests, one for ring & one for nested

**Comment by lpsinger** [comment]
> When it comes to writing the tests, do you have any suggestions for how to set it up? Right now I am thinking to somehow randomly generate an MOC (probably just by pulling a certain number of random UNIQ indices), decompose that into an l=29(?) healpix list (probably by seeing which l=29(?) healpix pixels intersect with said MOC), and then use that to get ranges I can run through the to_moc code to see if it produces the same MOC. Any thoughts there?
> 
> Oh and I will probably do 2 tests, one for ring & one for nested

Yup, just a simple round-trip test is all that you need here.

**Comment by bparazin** [comment]
Any recommended functions to find intersection of healpix pixels & MOC set? I'm having a bit of trouble efficiently breaking my randomly-generated MOC down into a list of healpix pixels of a given order

**Comment by bparazin** [comment]
Wait, I think the MOC just stores its values as l=29 healpix ranges? This might be easier than I though

**Comment by bparazin** [comment]
Actually, I have another idea. for a nested healpix pixel i, can you get the children in the next l up by doing i*4, i*4+1, i*4+2, i*4+3?

**Comment by bparazin** [comment]
This one is all working if you want to review it, I would like some eyes on how I do the level, since it's just set to l=12 as a max since higher than that caused a sigkill on my computer, and there's probably a better way to do it

File: healpix_alchemy/types.py [hunk 1]
```diff
@@ -28,6 +28,15 @@ def process_bind_param(self, value, dialect):
             value = int(value)
         return value
 
+    def to_moc(self, rangeSet, nside, index):
```
Comment (lpsinger) [inline healpix_alchemy/types.py:31]:
```suggestion
    @classmethod
    def to_moc(cls, rangeSet, nside, index):
```

File: healpix_alchemy/types.py [hunk 2]
```diff
@@ -28,6 +28,15 @@ def process_bind_param(self, value, dialect):
...
+        healpixSet = np.unique(np.concatenate([np.arange(start, stop, 1) for
+                                               (start, stop) in rangeSet]))
+        if index.lower() == "ring":
```
Comment (lpsinger) [inline healpix_alchemy/types.py:34]: Why do we need to support ring indexing at all? If we are trying to initialize a MOC object from an array of Tile instances, then the ranges represented by the Tiles will always use nested ordering.

File: healpix_alchemy/types.py [hunk 3]
```diff
@@ -28,6 +28,15 @@ def process_bind_param(self, value, dialect):
             value = int(value)
         return value
 
+    def to_moc(self, rangeSet, nside, index):
+        healpixSet = np.unique(np.concatenate([np.arange(start, stop, 1) for
+                                               (start, stop) in rangeSet]))
```
Comment (lpsinger) [inline healpix_alchemy/types.py:33]: This is generating a dense array of HEALPix pixel indices. This is very memory intensive. I think that `mocpy.MOC` can be initialized directly from an interval set, can't it?

File: healpix_alchemy/tests/test_types.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+from .. import types
+
+
+def random_MOC(max_l):
+    json = {str(level): list(
+        np.unique(
+            np.random.randint(0, 12 * (4 ** level),
+                              np.random.randint(12 * (4 ** level)))))
+            for level in range(max_l)}
```
Comment (lpsinger) [inline healpix_alchemy/tests/test_types.py:14]: Use the new-style numpy Generator API rather than the legacy global random state (see https://numpy.org/doc/stable/reference/random/index.html). Also, note that there is already a fixture in the benchmarks folder to generate a randomly subdivided sky map.

Rather than generate a random MOC, why not just generate a MOC for some simple known polygon or cone?

**Review by lpsinger — CHANGES_REQUESTED** [review]

File: healpix_alchemy/types.py (hunk 2 above)
Comment (bparazin) [reply healpix_alchemy/types.py:34]: I added it just for completeness, but have no issue with just using nested

File: healpix_alchemy/tests/test_types.py (hunk 4 above)
Comment (bparazin) [reply healpix_alchemy/tests/test_types.py:14]: I thought generating a random one is again more "complete" since it can cover wacky shapes or whatever, but I am again happy to just randomly select a cone of healpix pixels since the random selection is then just pick a center point and radius, which is much faster & easier

File: healpix_alchemy/types.py (hunk 3 above)
Comment (bparazin) [reply healpix_alchemy/types.py:34]: Ah, I missed that one at first since they don't explicitly say it, but it looks like you can pass a range in to the from_string method

**Comment by bparazin** [comment]
I am not using MOC's from_cone method here because it selects slightly different pixels as astropy.healpix's cone search, probably due to how it handles the edge cases. Now to fix the pep 8 errors

**Comment by bparazin** [comment]
Alright, good for another look-over

File: healpix_alchemy/tests/test_types.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    depth = 10
+    rng = np.random.default_rng()
+    lon = Longitude(360 * rng.random() * u.deg)
+    lat = Latitude((180 * rng.random() - 90) * u.deg)
+    radius = Angle(rng.normal(loc=10, scale=2.5) * u.deg)
+    hp = HEALPix(nside=2**depth, order='nested')
+    healpix_list = hp.cone_search_lonlat(lon, lat, radius)
+    depth_list = depth * np.ones(np.shape(healpix_list))
+    moc = MOC.from_healpix_cells(healpix_list, depth_list)
+    nested_hpx_ranges = [item for item in to_ranges(healpix_list)]
```
Comment (lpsinger) [inline healpix_alchemy/tests/test_types.py:29]: Replace with `MOC.from_cone`

File: healpix_alchemy/types.py [hunk 6]
```diff
@@ -28,6 +28,13 @@ def process_bind_param(self, value, dialect):
             value = int(value)
         return value
 
+    @classmethod
+    def to_moc(cls, rangeSet, nside):
```
Comment (lpsinger) [inline healpix_alchemy/types.py:32]: This class method should belong to the Tile class.

**Review by lpsinger — CHANGES_REQUESTED** [review]
Please provide a unit test that creates a MOC, inserts it into a database, retrieves it, and then converts it to back to a mocpy instance and checks that it survives the round-trip conversion.

**Comment by bparazin** [comment]
Ok! Probably some pep8 changes to make, but once I get those, this should be good for a re-review. I tried to make it more in-line with the other tests, so there's a much larger diff here

**Comment by mcoughlin** [comment]
@lpsinger Can you give feedback and/or merge here? I expect classes starting again will otherwise completely stall this.

**Review by lpsinger — COMMENTED** [review]
@mcoughlin, do you still need this? Sorry, I don't know why I failed to review it.

**Comment by mcoughlin** [comment]
Maybe we just close it :)
